core/process.py

```diff
@@ -1057,7 +1057,12 @@ def WhenStopped(self, stop_sig):
         # https://www.gnu.org/software/bash/manual/html_node/Exit-Status.html
         self.status = 128 + stop_sig
         self.state = job_state_e.Stopped
-        self.job_control.MaybeTakeTerminal()
+        # Only take back the terminal if we gave it to this process in the first
+        # place. Otherwise we might accidentally take it back from another
+        # process running in the foreground, causing it to get stopped with
+        # SIGTTIN/SIGTTOU.
```

Comment:
Hmm the bug makes sense

Does it make sense to set a boolean on this Process object if we called MaybeGiveTerminal in StartProcess()?

Reply:
Yeah, having a bit of state could be useful if you're worried about extra syscalls. I think it will need to be more dynamic than just at start, though (e.g. something repeatedly sent between the foreground and background might exhibit the same behavior)